### PR TITLE
expand model specification possibilities for structured populations 

### PR DESCRIPTION
**Pull request by jc-macdonald**
This pull request significantly expands and refactors the expression compilation and evaluation logic in the `op_system` module, making it more extensible, robust, and easier to maintain. The main improvements include broadening the set of allowed NumPy functions, introducing helper functions for state aggregation, centralizing validation and evaluation logic, and enhancing test coverage for new features and edge cases.

**Core logic improvements:**

* Expanded the set of allowed NumPy functions in expressions to include additional mathematical, trigonometric, and geometry-related functions such as `expm1`, `log2`, `log10`, `sin`, `cos`, `tan`, `sinh`, `cosh`, `tanh`, `hypot`, and `arctan2`. Also added a whitelist for helper functions (`sum_state`, `sum_prefix`).
* Refactored call validation logic into a dedicated `_validate_call` function, improving clarity and making it easier to extend allowed function sets in the future. [[1]](diffhunk://#diff-8e2086917cc8429224fcf95a737477ae0c947e9862470fcb26c6de9045a0ae71R233-R256) [[2]](diffhunk://#diff-8e2086917cc8429224fcf95a737477ae0c947e9862470fcb26c6de9045a0ae71L245-R281)
* Centralized state vector validation and equation evaluation into `_validate_state_vector` and `_evaluate_equations`, ensuring consistent error handling and reducing code duplication. [[1]](diffhunk://#diff-8e2086917cc8429224fcf95a737477ae0c947e9862470fcb26c6de9045a0ae71R375-R410) [[2]](diffhunk://#diff-8e2086917cc8429224fcf95a737477ae0c947e9862470fcb26c6de9045a0ae71L363-R457)

**New features:**

* Added support for reducer helper functions `sum_state` and `sum_prefix` in equations and aliases, allowing aggregation over state variables directly in user expressions.
* Updated the environment-building logic to inject these new helper functions for use in compiled expressions.

**Testing and configuration:**

* Added comprehensive tests for the expanded NumPy function whitelist, new helper functions, and stricter function call validation, ensuring correct behavior and preventing unauthorized function usage. [[1]](diffhunk://#diff-2fa3a022376673b04af8548e35d18cc87f0eac0a1985cffed7de0b40cf955447L151-R159) [[2]](diffhunk://#diff-2fa3a022376673b04af8548e35d18cc87f0eac0a1985cffed7de0b40cf955447R186-R268)
* Updated the `pyproject.toml` to ignore missing imports for the `flepimop2.typing` module in mypy, preventing unnecessary type checking errors during development.

See the [README.MD](https://github.com/ACCIDDA/op_system/tree/expand_specifcation_options) for updated API demonstration.

@shauntruelove I have added an asymmetric vaccination example to the README as requested, @pearsonca the ability to have flexible rate specification has also been added. Finally, I have removed the need to have [I1, I2, I3, ...] in both `State:` and `Chain:`.  Now `State` can simply be `State: [S, I, R]` and then just use the chain helper.

### CODE REVIEW

File: src/op_system/compile.py [hunk 1]
```diff
@@ -360,32 +419,42 @@ def _make_eval_fn(
...
-            except NameError as exc:
-                _raise_parameter_error(detail=f"unknown symbol in equation: {exc!s}")
-            except (ValueError, TypeError, ArithmeticError) as exc:
-                _raise_invalid_expression(detail=f"equation evaluation failed: {exc!r}")
-            out[i] = np.float64(val)
-
-        return out
+        return _evaluate_equations(eq_code=eq_code, env=env, n_state=n_state)
 
     return eval_fn
```
[the remark this answers is not in the document]
Comment (MacdonaldJoshuaCaleb) [reply src/op_system/compile.py]: agreed, I think a lot of this will be addressed by the various pedantic issues you opened @TimothyWillard
